app/models/user.rb

```diff
@@ -163,31 +163,37 @@ def course_student?
   end
 
   def nonvisitor?(course)
-    role(course) != CoursesUsers::Roles::VISITOR_ROLE
+    return true if admin?
+    !course_roles(course).empty?
   end
 
-  def role(course)
-    # If this is a new course, grant permissions.
-    return CoursesUsers::Roles::INSTRUCTOR_ROLE if course.nil?
-    # Give admins the instructor permissions.
-    return CoursesUsers::Roles::INSTRUCTOR_ROLE if admin?
+  # returns an array of roles a user has in a given course
+  def course_roles(course)
+    course.courses_users.where(user_id: id).order('role DESC').pluck(:role)
+  end
 
-    course_user = course.courses_users.where(user_id: id).order('role DESC').first
-    return course_user.role unless course_user.nil?
+  def highest_role(course)
+    roles = course_roles(course)
 
-    # User is in visitor role, if no other role found.
-    CoursesUsers::Roles::VISITOR_ROLE
+    return CoursesUsers::Roles::INSTRUCTOR_ROLE if admin?
+    return CoursesUsers::Roles::VISITOR_ROLE if roles.empty?
+
+    roles.first
   end
 
   EDITING_ROLES = [CoursesUsers::Roles::INSTRUCTOR_ROLE,
                    CoursesUsers::Roles::WIKI_ED_STAFF_ROLE].freeze
   def can_edit?(course)
     return true if admin?
-    return true if EDITING_ROLES.include? role(course)
+    return true if course_roles(course).any? { |role| EDITING_ROLES.include?(role) }
     return true if campaign_organizer?(course)
     false
   end
 
+  def can_view?(course)
```

Comment:
This is not needed, correct? (The tests can just be changed to `nonvisitor?`.)

Reply:
Update made and tested @ragesoss